packages/contracts/contracts/plugin/PluginSetupProcessor.sol

```diff
@@ -186,394 +247,394 @@ contract PluginSetupProcessor is DaoAuthorizable {
         });
     }
 
     /// @notice Applies the permissions of a prepared uninstallation to a DAO.
     /// @param _dao The address of the DAO.
-    /// @param _pluginSetup The address of the `PluginSetup` contract.
-    /// @param _pluginSetupRepo The repository storing the `PluginSetup` contracts of all versions of a plugin.
-    /// @param _plugin The address of the `Plugin` contract.
-    /// @param _currentHelpers The address array of all current helpers (contracts or EOAs) associated with the plugin that is uninstalled.
-    /// @param _permissions The list of multi-targeted permission operations to apply to the uninstalling DAO.
+    /// @param _dao The address of the installing DAO.
+    /// @param _params The struct containing the parameters for the `applyUninstallation` function.
     /// @dev The list of `_currentHelpers` has to be specified in the same order as they were returned from previous setups preparation steps (the latest `prepareInstallation` or `prepareUpdate` step that has happend) on which the uninstallation was prepared for.
-    function applyUninstallation(
-        address _dao,
-        address _plugin,
-        address _pluginSetup,
-        PluginRepo _pluginSetupRepo,
-        address[] calldata _currentHelpers, // TODO Isn't it sufficient to pass the helpers hash?
-        PermissionLib.ItemMultiTarget[] calldata _permissions
-    ) external canApply(_dao, APPLY_UNINSTALLATION_PERMISSION_ID) {
-        bytes32 setupId = _getSetupId(_dao, _pluginSetup, address(_pluginSetupRepo), _plugin);
-
-        // Check if the helpers match with those announced in the preparation step.
-        if (helpersHashes[setupId] != _getHelpersHash(_currentHelpers)) {
-            revert HelpersHashMismatch();
-        }
-
-        bytes32 storedPermissionsHash = uninstallPermissionHashes[setupId];
-        bytes32 passedPermissionsHash = _getPermissionsHash(_permissions);
-
-        // Check that the permissions match those announced in the preparation step.
-        if (storedPermissionsHash != passedPermissionsHash) {
-            revert PermissionsHashMismatch();
-        }
-
-        DAO(payable(_dao)).bulkOnMultiTarget(_permissions);
+    function applyUninstallation(address _dao, ApplyUninstall calldata _params)
+        external
+        canApply(_dao, APPLY_UNINSTALLATION_PERMISSION_ID)
+    {
+        bytes32 pluginId = _getPluginId(_dao, _params.plugin);
 
-        // Free up space by deleting the helpers and permission hash being not needed anymore.
-        delete helpersHashes[setupId];
-        delete uninstallPermissionHashes[setupId];
+        PluginState storage pluginState = states[pluginId];
 
-        emit UninstallationApplied({dao: _dao, plugin: _plugin});
-    }
+        bytes32 setupId = _getSetupId(
+            _params.pluginSetupRef,
+            pHash(_params.permissions),
+            bytes32(0),
+            bytes(""),
+            PreparationType.Uninstall
+        );
```

Comment:
Aren't we checking here the "theoretical" setupID vs the final one, when applying?

Effective permissions could have changed later on, but maybe it's not needed here